pyface/data_view/tests/test_data_view_widget.py

```diff
@@ -80,3 +89,243 @@ def test_header_visible_before_control(self):
+
+        self.assertEqual(self.widget.selection, [((1, 4), ()), ((2, 0), ())])
+        self.assertEqual(
+            self.widget._get_control_selection(),
+            [((1, 4), ()), ((2, 0), ())],
+        )
+
+    @unittest.skipIf(is_wx, "Selection type 'column' not supported")
+    def test_selection_type_column(self):
+        self.widget.selection_type = "column"
+        self._create_widget_control()
+
+        self.assertEqual(self.widget._get_control_selection_type(), "column")
+
+        self.widget.selection = [((0,), (2,)), ((1,), (4,))]
+
+        self.gui.process_events()
+
+        self.assertEqual(self.widget.selection, [((0,), (2,)), ((1,), (4,))])
+        self.assertEqual(
+            self.widget._get_control_selection(),
+            [((0,), (2,)), ((1,), (4,))]
+        )
+
+    @unittest.skipIf(is_wx, "Selection type 'item' not supported")
+    def test_selection_type_item(self):
+        self.widget.selection_type = "item"
+        self._create_widget_control()
+
+        self.assertEqual(self.widget._get_control_selection_type(), "item")
+
+        self.widget.selection = [((1, 4), (2,)), ((2, 0), (4,))]
+        self.gui.process_events()
+
+        self.assertEqual(
+            self.widget.selection,
+            [((1, 4), (2,)), ((2, 0), (4,))]
+        )
+        self.assertEqual(
+            self.widget._get_control_selection(),
+            [((1, 4), (2,)), ((2, 0), (4,))],
+        )
+
+    def test_selection_type_row_invalid_row_big(self):
+        self._create_widget_control()
+
+        with self.assertRaises(TraitError):
+            self.widget.selection = [((10,), ())]
+
+    def test_selection_type_row_invalid_row_long(self):
+        self._create_widget_control()
+
+        with self.assertRaises(TraitError):
+            self.widget.selection = [((1, 1, 1), ())]
+
+    def test_selection_type_row_invalid_column(self):
+        self._create_widget_control()
+
+        with self.assertRaises(TraitError):
+            self.widget.selection = [((1, 2), (2,))]
+
+    @unittest.skipIf(is_wx, "Selection type 'item' not supported")
+    def test_selection_type_item_invalid_row_too_big(self):
+        self.widget.selection_type = 'item'
+        self._create_widget_control()
+
+        with self.assertRaises(TraitError):
+            self.widget.selection = [((1, 10), (2,))]
+
+    @unittest.skipIf(is_wx, "Selection type 'item' not supported")
+    def test_selection_type_item_invalid_row_too_long(self):
+        self.widget.selection_type = 'item'
+        self._create_widget_control()
+
+        with self.assertRaises(TraitError):
+            self.widget.selection = [((1, 4, 5, 6), (2,))]
+
+    @unittest.skipIf(is_wx, "Selection type 'item' not supported")
+    def test_selection_type_item_invalid_column(self):
+        self.widget.selection_type = 'item'
+        self._create_widget_control()
+
+        with self.assertRaises(TraitError):
+            self.widget.selection = [((1, 2), (10,))]
+
```

Comment:
I am actually thinking about the code that consumes the `selection` list, rather than code that sets the list. But we do think to be concerned with both.

Developers who have to deal with the values in `widget.selection` will need to know whether they will receive row header items in the list too, otherwise their code might assume only values to do with leaf values and fail when a row header item is selected.

Given this affects user code consuming the content of `selection`, I still think this is a behaviour that should be tested.
At the very least, it should be documented.